### DISCUSSION TITLE
Custom help command

### DISCUSSION THREAD

**Discussion by s0r3-glitch** [nextcord/nextcord#252] [Q&A] [answered]
I was working on the implementation of a a help command when i ran into the problem of the command "help" being defined already. After further investigation I realized that  the help command is defined internally, however, I would like to either customize it or out right replace it since its currently just a code block of commands. Some things I would like to do are

- Add custom descriptions to commands and usage context when running `{prefix} help {command}`
- Make the whole thing an embed and possible add different pages that the user can scroll through with buttons
- Hide commands I don't want showing up like commands designed to only be ran by the bot owner
-  Just more general control of the command itself

I looked in the documentation of nextcord and found API refrences for help commands but it doesn't seem to be that well document on how to use them and only gives information like the classes attributes and methods. If someone could point me at an example of a custom help command I could look at or explain how to override the command that would be amazing. Thanks

**Comment by VincentRPS** [comment]
To disable the default help command put
help_command=None in your commands.Bot thing

**Reply by veni-vidi-code** [reply]
dont do that. i will send you some info in a sec

**Comment by veni-vidi-code** [comment, marked as answer]
From R. Danny (https://discord.com/channels/336642139381301249/559455534965850142/907961071330005052):
Create your own help command by subclassing HelpCommand or MinimalHelpCommand, then passing it to bot.help_command:

```python
class MyHelpCommand(commands.MinimalHelpCommand):
    def get_command_signature(self, command):
        return '{0.clean_prefix}{1.qualified_name} {1.signature}'.format(self, command)

class MyCog(commands.Cog):
    def __init__(self, bot):
        self._original_help_command = bot.help_command
        bot.help_command = MyHelpCommand()
        bot.help_command.cog = self

    def cog_unload(self):
        self.bot.help_command = self._original_help_command
```
or, you can set the help_command kwarg to None when instantiating you commands.Bot instance, and write your own.

Docs: https://nextcord.readthedocs.io/en/latest/ext/commands/api.html#help-commands

Migrating docs: https://nextcord.readthedocs.io/en/latest/migrating.html#helpformatter-and-help-command-changes

**Reply by VincentRPS** [reply]
Link the nextcord docs smh but good

**Reply by veni-vidi-code** [reply]
A more detailed explanation can be found at https://gist.github.com/InterStella0/b78488fb28cadf279dfd3164b9f0cf96

**Reply by s0r3-glitch** [reply]
Do I need to do this for every cog I have or should I just make a new cog and shove it all in there?

**Reply by veni-vidi-code** [reply]
one cog is enough

**Reply by s0r3-glitch** [reply]
ok thanks for you help and thanks for that gist link

**Reply by s0r3-glitch** [reply]
> A more detailed explanation can be found at https://gist.github.com/InterStella0/b78488fb28cadf279dfd3164b9f0cf96

I was reading this gist and was working on the implementation of the help command in a cog and I noticed in the code section for it that there was no setup function for it where it adds itself to the bot, am I safe to that its implied that I need to have this code snippet in it since its a cog?
`def setup(bot):
    bot.add_cog(cmds(bot))`
    
 I also noticed that there isnt a cmds class in the cog and am I also safe to assume that it doesnt need one since there isnt any special commands being made in the cog?

Edit: Just realized I am a complete idiot and that the class name doesnt need to be cmds and the only reason I was ever getting errors was because I never changed the setup function to reflect the class name change.

**Reply by DenverCoder1** [reply]
@s0r3-glitch You should have a `setup` function for every cog that is loaded including the Help cog.

In some examples it is included but in others it is left out, but in either case, it is still required.

**Reply by s0r3-glitch** [reply]
Ok thank you for the information!

**Reply by DenverCoder1** [reply]
In case it helps, here are some example help cogs that can be used for inspiration:

Embeds with Button pagination (with `nextcord-ext-menus` extension) - https://github.com/nextcord/nextcord-ext-menus/blob/master/examples/help_cog.py

Embeds with Cog Select Menus - https://github.com/DenverCoder1/Tutorial-Discord-Bot/tree/select-menu-help/modules/help

(I made a video tutorial for the select menu one which is the 6th and 7th videos in this playlist - https://www.youtube.com/playlist?list=PL9YUC9AZJGFG6larkQJYio_f0V-O1NRjy)
